llama_index/tools/ondemand_loader_tool.py

```diff
@@ -7,32 +7,35 @@
 
 from llama_index.tools.types import BaseTool, ToolMetadata
 from llama_index.readers.base import BaseReader
-from typing import Any, Optional, Dict, Type
+from typing import Any, Optional, Dict, Type, Callable, List
+from llama_index.readers.schema.base import Document
 from llama_index.indices.base import BaseIndex
 from llama_index.indices.vector_store import VectorStoreIndex
 from llama_index.tools.utils import create_schema_from_function
 from pydantic import BaseModel
+from llama_index.tools.function_tool import FunctionTool
 
 
 class OnDemandLoaderTool(BaseTool):
     """On-demand data loader tool.
 
-    Loads data with `reader.load_data()`, stores in index, and queries
-    for relevant data with a natural language query string.
+    Loads data with by calling the provided loader function,
+    stores in index, and queries for relevant data with a
+    natural language query string.
 
     """
 
     def __init__(
         self,
-        reader: BaseReader,
         index_cls: Type[BaseIndex],
         index_kwargs: Dict,
         metadata: ToolMetadata,
+        loader: Callable[..., List[Document]],
```

Comment:
nit: a bit weird that argument is now in 4th position as opposed to first. would probably prefer having it be in 1st position